pynitrokey/cli/nk3/test.py

```diff
@@ -0,0 +1,189 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2021 Nitrokey Developers
+#
+# Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+# http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+# http://opensource.org/licenses/MIT>, at your option. This file may not be
+# copied, modified, or distributed except according to those terms.
+
+import logging
+import sys
+from enum import Enum, auto, unique
+from hashlib import sha256
+from types import TracebackType
+from typing import Callable, Optional, Tuple, Type, Union
+
+from pynitrokey.helpers import local_print
+from pynitrokey.nk3.base import Nitrokey3Base
+from pynitrokey.nk3.device import Nitrokey3Device
+
+logger = logging.getLogger(__name__)
+
+
+TEST_CASES = []
+
+FIDO2_CERT_HASH = "ad8fd1d16f59104b9e06ef323cc03f777ed5303cd421a101c9cb00bb3fdf722d"
+
+
+ExcInfo = Tuple[Type[BaseException], BaseException, TracebackType]
+
+
+@unique
+class TestStatus(Enum):
+    SKIPPED = auto()
+    SUCCESS = auto()
+    FAILURE = auto()
+
+
+class TestResult:
+    def __init__(
+        self,
+        status: TestStatus,
+        data: Optional[str] = None,
+        exc_info: Union[ExcInfo, Tuple[None, None, None]] = (None, None, None),
+    ) -> None:
+        self.status = status
+        self.data = data
+        self.exc_info = exc_info
+
+
+TestCaseFn = Callable[[Nitrokey3Base], TestResult]
+
+
+class TestCase:
+    def __init__(self, name: str, fn: TestCaseFn) -> None:
+        self.name = name
+        self.fn = fn
+
+
+def test_case(name: str) -> Callable[[TestCaseFn], TestCaseFn]:
+    def decorator(func: Callable[[Nitrokey3Base], TestResult]) -> TestCaseFn:
+        TEST_CASES.append(TestCase(name, func))
+        return func
+
+    return decorator
+
+
+def log_devices() -> None:
+    from fido2.hid import CtapHidDevice
+
+    ctap_devices = [device for device in CtapHidDevice.list_devices()]
+    logger.info(f"Found {len(ctap_devices)} CTAPHID devices:")
+    for device in ctap_devices:
+        descriptor = device.descriptor
+        logger.info(f"- {descriptor.path} ({descriptor.vid:x}:{descriptor.pid:x})")
+
+
+@test_case("UUID query")
+def test_uuid_query(device: Nitrokey3Base) -> TestResult:
+    uuid = device.uuid()
+    uuid_str = f"{uuid:X}" if uuid else "[not supported]"
+    return TestResult(TestStatus.SUCCESS, uuid_str)
+
+
+@test_case("Firmware version query")
+def test_firmware_version_query(device: Nitrokey3Base) -> TestResult:
+    if not isinstance(device, Nitrokey3Device):
+        return TestResult(TestStatus.SKIPPED)
+    version = device.version()
+    return TestResult(TestStatus.SUCCESS, str(version))
+
+
+@test_case("FIDO2")
+def test_fido2(device: Nitrokey3Base) -> TestResult:
+    if not isinstance(device, Nitrokey3Device):
+        return TestResult(TestStatus.SKIPPED)
+
+    # Based on https://github.com/Yubico/python-fido2/blob/142587b3e698ca0e253c78d75758fda635cac51a/examples/credential.py
```

Comment:
Note: I think that if this is a verbatim copy, the whole copyrights remarks section have to be copied as well.

Reply:
The similarities are just dictated by the API, so I don’t think this is necessary here, but I can add the Yubico copyright statement to the file header to be sure.

Reply:
done